### ISSUE TITLE
Can't install loopback cli

### ISSUE THREAD

**Issue by danijel2017** [strongloop/loopback-cli#69]
Hi dear

I am new in loopback cli.
Now I am going to install cli of loopback but I got error and can't fix it.
Of course I searched google and stackoverflow but no answer there.
My os is win10
When install cli I got follow errors
http://prntscr.com/k7bhs1
Looking for your help.

Thanks

**Comment by dhmlau** [comment]
@danijel2017, have you installed `windows-build-tools`? i.e. `npm install -g windows-build-tools`.
Also, please see if this link can help you: 
https://stackoverflow.com/questions/44629411/node-js-npm-install-error

**Comment by danijel2017** [comment]
Of course I installed this but I can't to install loopback cli

**Comment by rafaelfragosom** [comment]
I'm having the same issue on a Mac.

Using yarn:
```
warning loopback-cli > generator-loopback > loopback-workspace > loopback-component-explorer > swagger-ui@2.2.10: No longer maintained, please upgrade to swagger-ui@3.
warning loopback-cli > generator-loopback > loopback-workspace > loopback > nodemailer@2.7.2: All versions below 4.0.1 of Nodemailer are deprecated. See https://nodemailer.com/status/
warning loopback-cli > generator-loopback > loopback-workspace > loopback > nodemailer > mailcomposer@4.0.1: This project is unmaintained
warning loopback-cli > generator-loopback > loopback-workspace > loopback > nodemailer > socks@1.1.9: If using 2.x branch, please upgrade to at least 2.1.6 to avoid a serious bug with socket data flow and an import issue introduced in 2.1.0
warning loopback-cli > generator-loopback > loopback-workspace > loopback > nodemailer > mailcomposer > buildmail@4.0.1: This project is unmaintained
[2/4] 🚚  Fetching packages...
error An unexpected error occurred: "https://registry.yarnpkg.com/deref/-/deref-0.7.4.tgz: Request failed \"404 Not Found\"".
info If you think this is a bug, please open a bug report with the information provided in "/Users/rafael/.config/yarn/global/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/global for documentation about this command.
```

Using npm:
```
npm WARN deprecated nodemailer@2.7.2: All versions below 4.0.1 of Nodemailer are deprecated. See https://nodemailer.com/status/
npm WARN deprecated mailcomposer@4.0.1: This project is unmaintained
npm WARN deprecated socks@1.1.9: If using 2.x branch, please upgrade to at least 2.1.6 to avoid a serious bug with socket data flow and an import issue introduced in 2.1.0
npm WARN deprecated buildmail@4.0.1: This project is unmaintained
npm WARN deprecated swagger-ui@2.2.10: No longer maintained, please upgrade to swagger-ui@3.
npm ERR! code E404
npm ERR! 404 Not Found: deref@https://registry.npmjs.org/deref/-/deref-0.7.4.tgz

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/rafael/.npm/_logs/2018-08-29T02_01_41_465Z-debug.log
```

Running node version 8.11.4

Here's the error part from the log file npm generated:
```
10215 verbose unlock done using /Users/rafael/.npm/_locks/staging-752146e6f61b7e27.lock for /Users/rafael/.nvm/versions/node/v8.11.4/lib/node_modules/.staging
10216 verbose type Error
10217 verbose stack Error: 404 Not Found: deref@https://registry.npmjs.org/deref/-/deref-0.7.4.tgz
10217 verbose stack     at fetch.then.res (/Users/rafael/.nvm/versions/node/v8.11.4/lib/node_modules/npm/node_modules/pacote/lib/fetchers/registry/fetch.js:42:19)
10217 verbose stack     at tryCatcher (/Users/rafael/.nvm/versions/node/v8.11.4/lib/node_modules/npm/node_modules/bluebird/js/release/util.js:16:23)
10217 verbose stack     at Promise._settlePromiseFromHandler (/Users/rafael/.nvm/versions/node/v8.11.4/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:512:31)
10217 verbose stack     at Promise._settlePromise (/Users/rafael/.nvm/versions/node/v8.11.4/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:569:18)
10217 verbose stack     at Promise._settlePromise0 (/Users/rafael/.nvm/versions/node/v8.11.4/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:614:10)
10217 verbose stack     at Promise._settlePromises (/Users/rafael/.nvm/versions/node/v8.11.4/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:693:18)
10217 verbose stack     at Async._drainQueue (/Users/rafael/.nvm/versions/node/v8.11.4/lib/node_modules/npm/node_modules/bluebird/js/release/async.js:133:16)
10217 verbose stack     at Async._drainQueues (/Users/rafael/.nvm/versions/node/v8.11.4/lib/node_modules/npm/node_modules/bluebird/js/release/async.js:143:10)
10217 verbose stack     at Immediate.Async.drainQueues (/Users/rafael/.nvm/versions/node/v8.11.4/lib/node_modules/npm/node_modules/bluebird/js/release/async.js:17:14)
10217 verbose stack     at runCallback (timers.js:810:20)
10217 verbose stack     at tryOnImmediate (timers.js:768:5)
10217 verbose stack     at processImmediate [as _immediateCallback] (timers.js:745:5)
10218 verbose cwd /Users/rafael/Development
10219 verbose Darwin 17.7.0
10220 verbose argv "/Users/rafael/.nvm/versions/node/v8.11.4/bin/node" "/Users/rafael/.nvm/versions/node/v8.11.4/bin/npm" "install" "-g" "loopback-cli"
10221 verbose node v8.11.4
10222 verbose npm  v5.6.0
10223 error code E404
10224 error 404 Not Found: deref@https://registry.npmjs.org/deref/-/deref-0.7.4.tgz
10225 verbose exit [ 1, true ]
```

**Comment by itsjef** [comment]
I'm also having the same issue. Running `npm i -g loopback-cli` gave:

```bash
npm WARN deprecated nodemailer@2.7.2: All versions below 4.0.1 of Nodemailer are deprecated. See https://nodemailer.com/status/
npm WARN deprecated mailcomposer@4.0.1: This project is unmaintained
npm WARN deprecated socks@1.1.9: If using 2.x branch, please upgrade to at least 2.1.6 to avoid a serious bug with socket data flow and an import issue introduced in 2.1.0
npm WARN deprecated buildmail@4.0.1: This project is unmaintained
npm WARN deprecated swagger-ui@2.2.10: No longer maintained, please upgrade to swagger-ui@3.
npm ERR! code E404
npm ERR! 404 Not Found: deref@^0.7.0

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/jef/.npm/_logs/2018-08-29T03_28_48_268Z-debug.log
```

**Comment by mrfelton** [comment]
Same issue. Currently impossible to install due to this.

**Comment by hacksparrow** [comment]
Looks like a problem in npm's registry. https://www.npmjs.com/package/deref says `0.7.4` is published, but it does it not exist in the download link.

**Comment by spali** [comment]
There is already an issue in the deref repo: json-schema-faker/deref#16
I could successfully work around by using version 3.1.0 of loopback-cli

**Comment by rafaelfragosom** [comment]
Thanks to @spali, the `yarn global add loopback-cli@3.1.0` works. 
Great workaround!

**Comment by mrfelton** [comment]
This seems to have been resolved now.

**Comment by joe-oli** [comment]
@mrfelton resolved? far from it. It is Oct 2018 and i attempted the 'Getting Started' instructions from the ibm/strongloop/loopback page at https://loopback.io/getting-started/

There is more involved than just saying... "Assuming you have already installed Node... Install the LoopBack CLI tool. "npm install -g loopback-cli"

So I have Windows 10, and latest node installed.
>npm -v
6.4.1
>node -v
v8.12.0

> npm install -g loopback-cli
==> results ==>
npm WARN deprecated nodemailer@2.7.2: All versions below 4.0.1 of Nodemailer are deprecated. See https://nodemailer.com/status/
npm WARN deprecated socks@1.1.9: If using 2.x branch, please upgrade to at least 2.1.6 to avoid a serious bug with socket data flow and an import issue introduced in 2.1.0
npm WARN deprecated mailcomposer@4.0.1: This project is unmaintained
npm WARN deprecated buildmail@4.0.1: This project is unmaintained
npm WARN deprecated swagger-ui@2.2.10: No longer maintained, please upgrade to swagger-ui@3.
npm WARN deprecated text-encoding@0.6.4: no longer maintained
C:\Users\jane.citizen\AppData\Roaming\npm\lb -> C:\Users\jane.citizen\AppData\Roaming\npm\node_modules\loopback-cli\bin\loopback-cli.js

> strong-ursa@0.10.1 install C:\Users\jane.citizen\AppData\Roaming\npm\node_modules\loopback-cli\node_modules\strong-ursa
> node-gyp rebuild


C:\Users\jane.citizen\AppData\Roaming\npm\node_modules\loopback-cli\node_modules\strong-ursa>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )
gyp ERR! configure error
gyp ERR! stack Error: Can't find Python executable "python", you can set the PYTHON env variable.
gyp ERR! stack     at PythonFinder.failNoPython (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\configure.js:484:19)
gyp ERR! stack     at PythonFinder.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\configure.js:509:16)
gyp ERR! stack     at C:\Program Files\nodejs\node_modules\npm\node_modules\graceful-fs\polyfills.js:284:29
gyp ERR! stack     at FSReqWrap.oncomplete (fs.js:152:21)
gyp ERR! System Windows_NT 10.0.17134
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\jane.citizen\AppData\Roaming\npm\node_modules\loopback-cli\node_modules\strong-ursa
gyp ERR! node -v v8.12.0
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok

> spawn-sync@1.0.15 postinstall C:\Users\jane.citizen\AppData\Roaming\npm\node_modules\loopback-cli\node_modules\spawn-sync
> node postinstall

npm WARN optional SKIPPING OPTIONAL DEPENDENCY: strong-ursa@0.10.1 (node_modules\loopback-cli\node_modules\strong-ursa):
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: strong-ursa@0.10.1 install: `node-gyp rebuild`
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: Exit status 1

+ loopback-cli@4.2.1
added 858 packages from 954 contributors in 90.74s

------------- END OF Getting Started attempt ------------------

It's clear the 'getting started' page needs special instructions for Windows users, such as:

Python v2.7.x series must be installed + configured.
OpenSSL- for windows must be(?) installed + configured
Visual Studio Build Tools must be (?) installed + configured.

The npm package itself should be updated to replace Deprecated dependencies with newer versions.

**Comment by raymondfeng** [comment]
`strong-ursa` is an optional dependency and the error messages are just warnings. The installation of `loopback-cli` should not be impacted by that.

**Comment by vivaldus** [comment]
@raymondfeng "The installation of loopback-cli should not be impacted by that" -- this problem causes the build to fail. Very frustrating for users who want to dive straight in.

**Comment by methodbox** [comment]
I've resolved this issue by doing a few simple things to take control of the loopback dependencies:

1. Be sure node-gyp is installed globally. npm call sometimes fall back to this when it's missing from a module dependency. May not be necessary. ```sudo npm install -g node-gyp```.

2. Assuming your macOS user is an admin and not a restricted user, give the Admins permission to the node_modules folder for the system. You'll need to use Finder and navigate to ```/opt/local/lib/node_modules/```. Ctrl-Click/right-click and select Get Info. Click the lock on the bottom-right and authorize. Click the "+" on the bottom-left. Select Administrators. You don't need to add your user, because your user is an admin and sudo will allow you to modify this folder. Click the lock to finish.

3. Loopback at ```/opt/local/lib/node_modules/loopback``` on my system was owned by root:wheel; this shouldn't be as npm is installed with sudo and the npm folder is actually owned by user:wheel, as it should be; this was preventing writing to the folders within loopback.  Chowning this folder is safe as long as you maintain the group ```wheel``` as it will still require sudo/admin to access.  
```chown -R username:wheel opt/local/lib/node_modules/loopback-cli```

4. Loopback never installed ```strong-ursa```; the folder was missing from ```opt/local/lib/node_modules/loopback-cli/node_modules```. Change into that directory and install it for loopback:  ```cd opt/local/lib/node_modules/loopback-cli```.  YOU MUST be in that directory because we are installing local to loopback. ```npm install strong-ursa```.

5.  Now, so that loopback recognizes it's dependencies (it does a node-gyp rebuild at the end, which is where it errors, so your install isn't complete yet), just run the loopback install again: ```sudo npm install -g loopback-cli```

Everything should install w/out errors.

**Comment by methodbox** [comment]
@joe-oli 
Install python using the 2.7 installer from their site.
Install OpenSSL 1.0.2q 64bit from the installer here:  https://slproweb.com/products/Win32OpenSSL.html  - you want the one labeled "Win 64 OpenSSL 1.0.2q" - this is the issue that is fixed by Strongloop's own fork of ```strong-ursa``` here: https://github.com/strongloop-forks/ursa but doesn't appear to have been backported into Loopback v3 yet.

```npm install -g windows-build-tools```  <-- do this in powershell as admin.
You should now be able to install w/out errors.

I've tried this on Windows 2016.

**Comment by phanducquang** [comment]
> @joe-oli
> Install python using the 2.7 installer from their site.
> Install OpenSSL 1.0.2q 64bit from the installer here: https://slproweb.com/products/Win32OpenSSL.html - you want the one labeled "Win 64 OpenSSL 1.0.2q" - this is the issue that is fixed by Strongloop's own fork of `strong-ursa` here: https://github.com/strongloop-forks/ursa but doesn't appear to have been backported into Loopback v3 yet.
> 
> `npm install -g windows-build-tools` <-- do this in powershell as admin.
> You should now be able to install w/out errors.
> 
> I've tried this on Windows 2016.

its works for me

**Comment by Nanuim** [comment]
for me... I needed to install python first... thanks!

> @joe-oli
> Install python using the 2.7 installer from their site.
> Install OpenSSL 1.0.2q 64bit from the installer here: https://slproweb.com/products/Win32OpenSSL.html - you want the one labeled "Win 64 OpenSSL 1.0.2q" - this is the issue that is fixed by Strongloop's own fork of `strong-ursa` here: https://github.com/strongloop-forks/ursa but doesn't appear to have been backported into Loopback v3 yet.
> 
> `npm install -g windows-build-tools` <-- do this in powershell as admin.
> You should now be able to install w/out errors.
> 
> I've tried this on Windows 2016.

for me... I needed to install python first... thanks!

**Comment by methodbox** [comment]
Glad that fix is working for you guys, but I’m surprised that Loopback still hasn’t backported their `strong-ursa` fix...

**Comment by mehransahandi** [comment]
Do as follow !

sudo chown -R $(whoami) ~/.npm
sudo chown -R $(whoami) /usr/local/lib
sudo chown -R $(whoami) /usr/local/bin

**Comment by dr3x** [comment]
I had one of the issues reported by the OP @danijel2017 - the missing libeay32.lib:
`Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.`
  `ursaNative.cc`
  `win_delay_load_hook.cc`
`LINK : fatal error LNK1181: cannot open input file 'C:\OpenSSL-Win64\lib\libeay32.lib'  `

Looks like in recent versions that lib has been renamed, so I installed OpenSSL and then put in a symlink to the lib with the new name in the folder structure it was expecting:

`C:\OpenSSL-Win64\lib>mklink .\libeay32.lib "C:\Program Files\OpenSSL-Win64\lib\libcrypto.lib" `
`symbolic link created for .\libeay32.lib <<===>> C:\Program 'Files\OpenSSL-Win64\lib\libcrypto.lib'`

After this workaround loopback-cli installed without error.  
So it appears to be due to breaking changes in OpenSSL.

**Comment by methodbox** [comment]
I posted the likely cause of this months ago (the `strong-ursa` fix that hadn’t been backported and still hasn’t I don’t think) - is this repo even being maintained?

There’s only 14 open issues and this is the second most recent with no response from the repo owners.

@strongloop @dhmlau are you still maintaining this? Do you not respond to issues? Or is there a better place to report bugs?

**Comment by dhmlau** [comment]
Discussed with @raymondfeng a bit on this... He mentioned that ursa requires openssl.  On Windows, openssl has to be installed manually.  Since Node.js 10 now has rsa support, there's a possibility that we get rid of ursa dependencies. 

@raymondfeng, what do you think it's a good way to resolve this?  Thanks.

**Comment by dhmlau** [comment]
Cross posting @raymondfeng's comment from https://github.com/strongloop/loopback-cli/issues/87#issuecomment-498114452:

> We're trying to remove ursa dependency - see strongloop/strong-soap#224
> We just released strong-soap@1.20.0. Can you try run npm i -g loopback-cli again?

**Comment by dhmlau** [comment]
Closing it as resolved.  @danijel2017 and others on this thread, if you still encounter issues in installing loopback-cli, please open a new issue. Thanks!
